File: SIC_Backend/Repositories/UsersRepository.cs
using Microsoft.EntityFrameworkCore;
using SIC_Backend.Data;
using SIC_Backend.Data.DTOs;
using SIC_Backend.Data.Models;

namespace SIC_Backend.Repositories
{
    public class UsersRepository(ApplicationDbContext dbContext) : IUsersRepository
    {
        public UserDTO GetUserById(string id)
        {
            var user = dbContext.Users.Where(u => u.Id == id)
                .Include(u => u.Places)
                .ThenInclude(p => p.Sections)
                .ThenInclude(s => s.Devices)
                .Select(UserDTO.FromUser)
                .First();

            return user;
        }

        public UserDTO GetUserByUsername(string username)
        {
            var user = dbContext.Users.Where(u => u.UserName == username)
                .Include(u => u.Places)
                .ThenInclude(p => p.Sections)
                .ThenInclude(s => s.Devices)
                .Select(UserDTO.FromUser)
                .First();

            return user;
        }

        public async Task<bool> IsUserNameAvailableAsync(string username)
        {
            return await dbContext.Users.AllAsync(u => u.UserName != username);
        }

        public async Task<IEnumerable<NotificationToken>> GetNotificationTokens(string userId)
        {
            return await dbContext.NotificationTokens.Where(nt => nt.UserId == userId).ToListAsync();
        }
    }
}

File: sic_home/lib/repositories/users_repository.dart
import 'package:dio/dio.dart';
import 'package:sic_home/apis/users_api.dart';
import 'package:sic_home/models/authentication/notification_token.dart';
import 'package:sic_home/models/user.dart';
import 'package:sic_home/repositories/interfaces/i_users_repository.dart';

class UsersRepository implements IUsersRepository {
  final UsersApi _usersApi = UsersApi(Dio());

  @override
  Future<bool> isUserNameAvailable(String userName) async {
    return await _usersApi.isUserNameAvailable(userName);
  }

  @override
  Future<List<NotificationToken>> getNotificationTokens(String userId) async {
    return await _usersApi.getNotificationTokens(userId);
  }

  @override
  Future<User> getUserByUserName(String userName) async {
    return await _usersApi.getUserByUserName(userName);
  }

  @override
  Future<User> getUserById(int id) async {
    return await _usersApi.getUser(id);
  }
}
